File: 2024/day_2.py
from itertools import pairwise
from operator import sub
from typing import cast



def report_is_safe(report: list[int]) -> tuple[bool, int]:
    """
    Checks if a report is safe according to the rules of the challenge and returns
    the state and the index of the first noted issue with the report
    """
    pairs = list(pairwise(report))
    diffs = [cast(int, sub(a, b)) for a, b in pairs]
    
    contains_neg = False
    contains_pos = False
    for index, value in enumerate(diffs, start=1):
        if abs(value) > 3 or abs(value) < 1:
            return (False, index)
        
        if value >= 0: 
            contains_pos = True
        else: 
            contains_neg = True
        
        if contains_neg and contains_pos:
            return (False, index)
    
    return (True, -1)
    

def report_is_safe_with_dampener(report: list[int]) -> tuple[bool, int]:
    """
    This runs the safety check function and ignores a single case of safety concern
    """
    is_safe, failure_index = report_is_safe(report)
    if not is_safe:
        if failure_index == 1:
            for index in range(0, 2):
                record = report.pop(index)
                result = report_is_safe(report)
                if result[0]:
                    return (True, index)
                report.insert(index, record)
            return (False, 1)
        else:
            report.pop(failure_index)
            return report_is_safe(report)
    return (True, -1)


if __name__ == "__main__":
    # reports = open("2024/data/day_2_input.txt").readlines()
    reports = [
        # [7, 6, 4, 2, 1],
        # [1, 2, 7, 8, 9],
        # [9, 7, 6, 2, 1],
        # [1, 3, 2, 4, 5],
        # [8, 6, 4, 4, 1],
        # [1, 2, 3, 7, 6],
        # [5, 7, 9, 12, 13, 11],
        # [10, 9, 7, 4, 4, 1],
        [1, 6, 7, 8, 9],
        [3, 10, 11, 12, 13],
        [8, 6, 4, 4, 1],
        [2, 4, 6, 5, 7],
        [9, 8, 7, 3, 1],
    ]
    safe_report_count = 0
    for report in reports:
        report = [int(item) for item in report]
        result = report_is_safe_with_dampener(report) 
        if result[0]:
            safe_report_count += 1
            # print(f"{report = }, {result = }")
        else:
            print(f"{report = }, {result = }")
            
    print(f"Safe Report Count = {safe_report_count}")

File: 2024/day_2.rs
use std::fs;


fn pairwise<T: Clone>(sequence: Vec<T>) -> Vec<(T, T)> {
    let mut pairs = Vec::new();
    
    for (index, value) in sequence.iter().enumerate() {
        if index == (sequence.len() - 1) {
            break
        }
        
        let pair = (value.clone(), sequence[index + 1].clone());
        pairs.push(pair);
    }
    
    pairs
}

fn report_is_safe(report: Vec<i32>) -> (bool, Option<usize>, Option<String>) {
    let pairs = pairwise(report);    
    let diffs = pairs.iter().map(|(a, b)| b - a).collect::<Vec<_>>();
    
    let mut contains_neg = false;
    let mut contains_pos = false;
    
    for (index, value) in diffs.iter().enumerate() {
        
        if value.abs() > 3 || value.abs() < 1 {
            return (false, Some(index), None);
        }
        
        if *value > 0 {
            contains_pos = true;
        } else {
            contains_neg = true;
        }
        
        if contains_neg && contains_pos {
            return (false, Some(index), None);
        }
    }
    
    (true, None, None)
    
}

fn report_is_safe_with_dampener(report: Vec<i32>) -> (bool, Option<usize>, Option<String>) {
    let mut report_copy = report.clone();
    let (is_safe, failure_index, _) = report_is_safe(report_copy.clone());
    
    if !is_safe {
        if (0..3).contains(&failure_index.unwrap()) { 
            for index in 0..3 {
                let faulty_record = report_copy.remove(index);
                let (second_attempt, _, _) = report_is_safe(report_copy.clone());
                
                if second_attempt {
                    return (true, Some(index), None);
                }
                report_copy.insert(index, faulty_record);
            }
            return (false, Some(1), None);
        } else {
            report_copy.remove(failure_index.unwrap());
            return report_is_safe(report_copy.clone());
        }  
    }
    
    
    (true, None, None)
}

fn get_total_safe_report(reports: Vec<Vec<i32>>, with_dampener: bool) -> i32 {
    let mut total = 0;
    let report_safety_func = if with_dampener { report_is_safe_with_dampener } else { report_is_safe };
    
    for report in reports {
        let (is_safe, _, _) = report_safety_func(report);
        
        if is_safe {
            total += 1;
        }   
    }
    
    total
}


fn main() {
    let content = fs::read_to_string("2024/data/day_2_input.txt").expect("Failed to Read Input Data");
    let reports = content.lines()
            .map(
                |line: &str| 
                    line.split_whitespace()
                        .map(|item| item.parse::<i32>().expect("failed to convert to Integer"))
                        .collect::<Vec<_>>()

            ).collect::<Vec<_>>();
    
    // println!("Reports: {reports:?}");
    
    let total = get_total_safe_report(reports, false);
    println!("Total: {total}");
}